### ISSUE TITLE
Hotkeys to dim your display

### ISSUE THREAD

**Issue by heywolf15** [Tyrrrz/LightBulb#141]
Slightly slower dimming with (Alt-PgDn, Alt-PgUp)

**Comment by Tyrrrz** [comment]
Can you clarify why you would want that?

**Comment by heywolf15** [comment]
Sorry but, isn't it pretty clear?
To change the brightness quickly, no need to open the app

**Comment by xcasxcursex** [comment]
Now I'm curious too, why would you want that? Is that for those times you wake up at mid-day with a hangover and want the screen to act like it's nighttime? ;)
Totally not asking because I just woke up at mid-day. Honest XD

**Comment by heywolf15** [comment]
Why is this a weird thing lol.

It's just a hotkeys, so you don't have to open the app
One of the reason I use other competitor, Flux

**Comment by xcasxcursex** [comment]
Well it's weird because the idea of this kind of software is to have your monitor match the daylight, for health reasons (because our body is tuned to react to natural light and the artificial monitor light being so different to the sunlight can mess with you, such as making it hard to get to sleep for example) so it seems strange to have the monitor match some other lighting condition. You shouldn't need to open the app to change it arbitrarily because you shouldn't need to change it arbitrarily.

I'm not arguing that you have some other need , aside from the intention of the software (either flux or lightbulb, both exist for the same reason), I just don't understand what it would be. I'm just curious. If you're going to ask "why is this a weird thing" I guess the correct response is "How is it NOT a weird thing? Also.... Why don't you just tell us why you'd want to even do this?" I mean.... There must be some good reason why you do it or you wouldn't be asking, right? So what is it? I'm just asking is all :)

**Comment by heywolf15** [comment]
I think what I say is not clear

Hotkeys for the brightness, dimming the brightness

**Comment by xcasxcursex** [comment]
Ahhh I see what you mean now sorry I misunderstood! You did say "dim" and "brightness". I guess I just kinda think of those as a secondary feature to an app like this so it didn't 'click' for me.

I guess I can see where that might be more useful... Like right now I have the room light on but if it were off it would be nice to have less monitor brightness.... On my TV there is an automatic sensor to alter brightness according to ambient light, and I know some computer monitors (especially laptops) have something similar, so I understand how it can be useful to alter brightness by itself without changing gamma.... And yeh, easier to get to it with shortcut keys, you wouldn't want to have to open the app.

It's not the first time this kind of thing has been asked either....
There are some other issues related to this too:
https://github.com/Tyrrrz/LightBulb/issues/19
https://github.com/Tyrrrz/LightBulb/issues/28
https://github.com/Tyrrrz/LightBulb/issues/126
https://github.com/Tyrrrz/LightBulb/issues/110


So really this could be a request for a few things:

Super-supreme version:
Implement separate brightness offset from gamma offset (so they don't always change at the same time as they do now)
Optionally link the two offsets (linked as they are now, or in the new behaviour, brightness would be separate)
Option to offset brightness from time-defined curve (brighter or darker than normal curve like we have now), or, Explicitly set brightness (just set the brightness to X regardless of the time of day)
Keybinds to control offset/explicit setting, up and down

Does that sound about right?

Simple version:
In its simplest form, I guess it could simply be a variable, controlled by two keybinds, which offsets the brightness from whatever the app normally would set. 

That seems much simpler to implement but I'm not sure if it would fit the requirement? 

I wonder about whether this is 
a) Practical - is it too complex to code? I have lots of time to donate right now if you would like me to try, @Tyrrrz ... but I don't know the codebase well, so I don't know if I would basically be doing another re-work or if it would just be some modifications to what is already there. Your answer to #110 makes me wonder about this, but was that before the 2.0 re-work? Would the "simple version" above, be more practical?
b) Performant - one of the biggest reasons to use Lightbulb rather than f.lux is that it performs so much better (especially for gamers). If all these new features could slow it down/place significant load on the system, etc, that would be bad.

**Comment by Tyrrrz** [comment]
To be honest, I don't want to add anything that could make the existing gamma flow more complex, unless it's really warranted.

Original version of LightBulb expressed the gamma flow as a state machine with only 3 variables: current time, day temperature, night temperature. As the software evolved, the number of variables has grown and, as a result, the number of state permutations expanded as well. It grew so much that it became unwieldy and there were often race conditions because the variables often changed separately from one another as a result of different external factors (multiple timers, system events, user actions, etc).

When reworking it to v2.0, I took a bit different approach. The state is expressed reactively by re-evaluating all of the variables whenever they change (think redux, but simple). The variables themselves still come from different sources, but they are mostly reduced into a smaller number of decisions that end up influencing the state. For example, there is a single decision for changing gamma, for pausing, for invalidation, etc. There are no race conditions as a result of it, but it requires a lot of planning. In the current version of LightBulb, the state is evaluated based on over 30-40 different variables, resulting in potentially enormous number of valid states. The fact that LightBulb works, despite a bunch of different options you can set in various combinations to influence its workflow, other applications changing or invalidating gamma, Windows caveats, multithreading etc. is not just luck. :)

Anyway, my point is that introducing a change into all of this is not trivial and has to be worth it. At this point, I'm still not convinced that additional manual adjustment of brightness is worth it. I, personally, don't find myself ever needing to use it. In contrast, automatic brightness adjustment is very useful on my desktop PC as it doesn't have it built-in.

**Comment by xcasxcursex** [comment]
Well I feel I should say; as I mentioned, my TV, which is my main screen (although my "Primary" in the OS is my GSync monitor I use that for gaming and not for general use) has a brightness sensor which reduces brightness when ambient light in the room is lower/increases brightness when ambient light is increased.... So I personally would never have a need for this manual adjustment, because my TV is doing it automatically..... However, I can see that it could be useful. For people who do not have this feature in their display, it would be nice to be able to turn down the brightness when they turn down the lights. 

I'm also reminded of a custom userscript I wrote, that fades video playback in my browser, so that if I decide to have a nap with a movie or youtube playing in the background, the bright light doesn't bother me. Again, I myself already have a tool for this, but again it makes me see how manual brightness control could have good cause for practical application.

So, I personally would not have a strong need for manual brightness control, but I can see how it would be very useful for others.... and honestly, if it was already in LightBulb, I would certainly be using it, and would not be bothering to write custom userscripts for my browser. 

And obviously, from the previous requests for this, and from the requests for it and eventual inclusion in similar software (f.lux), it's not entirely uncommon for people to see some use for it.

In short: I think the idea has value.

The trouble is, implementing it in such a way that is practical and performant. As you say @Tyrrrz , it has to be "worth it", and I think that there is "worth", but the definition of "it" is not really clear to us mere mortals because you are the only expert here :) 

I understand that if we are talking about a state machine then it is not just (+1) variable, it is x^(y(+1)) states.... Which would be impractical and possibly poorly performing, and definitely not "worth it".... That being said, I think if we were to look at the "Simple version" I posted above, then it would not be so complex. As far as I can see, brightness is stored as part of the ColorConfiguration, and nothing would need to be changed about that. Everything can behave as normal, right down to when the SetGamma() is called. Then instead of
`
    ....
                    deviceContext.SetGamma(
                        GetRed() * colorConfiguration.Brightness,
    ....
`

Simply add another multiplication there, with the new variable. which is controlled by hotkeys. and a simple function to limit the value to sensible ranges when hotkeys are pressed.

So like:
`
    ....
                    deviceContext.SetGamma(
                        GetRed() * colorConfiguration.Brightness * colorConfiguration.ManualBrightness,
    ....
`. which is controlled by hotkeys. and a simple function to limit the value to sensible ranges when hotkeys are pressed.

**Comment by Tyrrrz** [comment]
Sure. Let be get back to it when I have more time.

**Comment by xcasxcursex** [comment]
Would you like me to try and start on this for you (in a fork/PR)? I would like to help, but I also don't want to rush you or get in the way :) ... And I might just do it totally wrong lol

**Comment by Tyrrrz** [comment]
I would hold off on this a bit. I have some plans for refactoring in the near future so it would be better to get those in first.

**Comment by AnRetset** [comment]
Been a while since this thread had activity - just wanted to pop in and say that around half of my evenings I hop into a game or two, and will have to go into Lightbulb settings and manually increase my evening brightness by around 15% during gameplay and my colour temperature by 1000K to be able to subtly make out blues and details; and then I go back and manually reduce them to their original values when I'm done playing.  So for me, it would be quite lovely to have a more accessible way of overriding Lightbulb's current brightness value.  I too am used to Alt+PageUp/Alt+PageDown from f.lux.

If @xcasxcursex's suggestion above does not add too much complexity, then I would also propose a colour temperature override value.  If that feels like it would bog the pipeline down too much, then perhaps as an alternative we can have the option to select from a list of presets?

For example, my "gaming" preset could be the same values for daytime, and then just slightly higher brightness and temperature than what my "default" preset is for evening.  And then we could change presets with hotkeys without adding any additional variables to the state machine, if I understand it correctly.

That said, manual control over brightness (and temperature) overrides/offsets via hotkeys would really be preferred, as it makes more sense to have granular control over something than be limited to what you've set up in presets.  Hope my use case and explanation make a bit of sense here.

**Comment by fabiorfreitas** [comment]
Hi, I'd like to add that while the more complex suggested implementations are definitely wanted, we could also use simple hotkeys for dialing up or down the brightness settings as they are currently.

For me, the main use would be 21:00 is as dark outside as 03:30, but I could use different brightness levels for them, even if manually setting with hotkeys (I already do that from function keys on my laptop, but it's not enough).

**Comment by anderrom** [comment]
I would like this option aswell:
https://imgur.com/a/kyB9ID4
These are from F.lux, very useful in my opinion.

**Comment by Tyrrrz** [comment]
Let's consider some design challenges.

**Option 1: hotkeys that change settings** (as suggested by @fabiorfreitas)

This seems to be the simplest from implementation perspective but they are things to consider when it comes to UX:

1. When you change temperature/brightness settings, should they be automatically saved? Or not?
2. If not, when you open up the settings dialog, should these settings be reset to last saved state or kept at the adjusted state? If latter, that may be confusing. If former, I'm not sure if that's the desired behavior.

Also, which exact temperature/brightness should be changed when through these hotkeys? Let's assume the user pressed "increase brightness" mid-day, in that case it seems logical to change the day-time brightness. If it's during sunrise though, what was the intent? Day-time or night-time?

**Option 2: separate offset for temperature and brightness** (as suggested by @xcasxcursex)

Should the offset be linear or multiplicative? It seems in your code example it's multiplicative, but I'm not sure if that's desired. With multiplicative offset, different actual values of brightness/temperature will be affected in different ways.

How should the offset be presented to the user? I'd assume some UI element on the main view showing that the brightness/temperature has been manually adjusted, with the option to reset it.

Additionally, what should happen to manual offset if the user opens settings and starts moving the actual brightness/temperature sliders around? I'd guess the offset should be reset then, but that would add complexity.

**Comment by anderrom** [comment]
Well, I think you should keep it simple...
So for option 1, yes, when you change the temperature with the hotkeys they are automatically saved, this is how I expected the application to behave.
You don't need to add a slider or a UI option to change, since it is an advanced option that very few users will need, you can only add it as a hotkey option, so it is up to the user to bind and use this option. Just add some kind of feedback when the user change brightness with the hotkey, like a message overlay or simple displaying the current brightness next to the temperature. That's enough for me.

By default the brightness is 100% at all times, if the user changes it, it stays that way until the user changes again. No changes according to time.

**Comment by fabiorfreitas** [comment]
Regarding my previous input, I was referring to actually triggering the settings in the same way we would with the UI. As so, they would be permanent. The only extra "direction" would be to enact those changes to the current settings. Perhaps the hotkeys would have two sets of codes (for daytime and nighttime) and the same mechanism that makes the transition would also change the hotkeys' behavior each time.

That's supposed to be the quick way, (hopefully) less exhausting for you.

In the long therm, I believe this whole feature should be treated as an override (reviewing my text, perhaps "offset" is better). It would not mess with the actual settings, but simply add to or subtract from the final values before the are applied.

This implementation, if possible, would also be great in not disrupting transitions, as overrides wouldn't mess with daytime, nor nighttime settings.

If you think there should be something on the UI, I believe no more than a simple informational, non-interactive box would suffice. Something as:

**Overrides:** +300-K | -22% Brightness
 
Overlays, perhaps taking advantage of Windows' native notifications, also would look very good!

Other things to consider would be regarding how permanent these overrides would be. There are a lot of equally reasonable approaches:

1. Overrides remain until user sets them back (optionally with a reset hotkey);
2. Overrides remain active until LightBulb is closed;
3. Overrides last for a given amount of time (I don't think this is a really good idea, but it's still worth noting);
4. Overrides are ended by the next transition (I find this concept interesting, but perhaps problematic).

**Comment by Tyrrrz** [comment]
@anderrom 

> By default the brightness is 100% at all times, if the user changes it, it stays that way until the user changes again. No changes according to time.

Didn't really understand this part. Let me rephrase, if the current state is in transition to day (sunrise), which setting should "increase brightness" change? Night-time brightness or day-time brightness? Or both?

@fabiorfreitas 

> Regarding my previous input, I was referring to actually triggering the settings in the same way we would with the UI. As so, they would be permanent. 

Yes, but on the UI you also need to press the "Save" button to persist the changes. I assume you want that to happen automatically?

> Perhaps the hotkeys would have two sets of codes (for daytime and nighttime) and the same mechanism that makes the transition would also change the hotkeys' behavior each time.

Not sure I understood this. Which behavior should the hotkeys have if it's currently night-to-day transition, i.e. sunrise?

> In the long therm, I believe this whole feature should be treated as an override (reviewing my text, perhaps "offset" is better). It would not mess with the actual settings, but simply add to or subtract from the final values before the are applied.

I think both approaches are more or less equal in terms of complexity, so we might as well target the "long term" variant. I've listed some considerations regarding it in the previous post, where it was referred to as "option 2".

> Other things to consider would be regarding how permanent these overrides would be. There are a lot of equally reasonable approaches:

This is one of the open questions. I think 1 is a must, 2 is going to be implicit assuming we're going with "option 2" and not "option 1". 3 and 4 seem like a stretch. I'm also consider what should happen when the user opens the settings dialog and changes the current settings, should the offset stay or get reset?

**Comment by anderrom** [comment]
I was thinking in a independent brightness settings, no need to have nighttime brightness or daytime brightness, so it would be much simpler to implement, and that is all usability I would need from this.

**Comment by xcasxcursex** [comment]
> Let's consider some design challenges.

This is always complex...here we go! :)
First: a TL;DR:

This issue is intrinsically linked to #131 . That requires a refactor, and if it's done right, so will this... The app is growing beyond it's original design. I think it will be important to consider this issue, when approaching that one. 

In the meantime, i think that an overly-simple approach to this issue, could stand as a stop-gap temporary solution until more flexible methods can be done in conjunction with #131 

The most simple approach I can imagine, would be the following:

- Add three hotkeys to the existing hotkeys dialog: Increase/Decrease/Reset brightness offset
- Have these keys modify a new variable stored in the existing colorConfig class
- Use this variable as a multiplier in setGamma calls as per above.

- Optional: Add a slider to the UI to display/set the variable




> **Option 1: hotkeys that change settings**
>     1. When you change temperature/brightness settings, should they be automatically saved? Or not?

Optional. For me, if I restarted my PC I'd expect things to return to normal.... but we can already see, that's not for everybody.


>     2. If not, when you open up the settings dialog, should these settings be reset to last saved state or kept at the adjusted state? If latter, that may be confusing. If former, I'm not sure if that's the desired behavior.

I can see a need for both. For example if I am modifying the nighttime gamma, I will want to know what it looks like under default brightness but also my modified brightness.

> Also, which exact temperature/brightness should be changed when adjusted through these hotkeys? Let's assume the user pressed "increase brightness" mid-day, in that case it seems logical to change the day-time brightness. If it's during sunrise though, what was the intent? Day-time or night-time?
> 
> **Option 2: separate offset for temperature and brightness** (as suggested by @xcasxcursex)

I should clarify my post here: I was intending to point out that there are many ways to approach a problem like this - it is easy to say "hotkeys to dim your display", but what does that actually mean, programmatically?

In order to demonstrate this, I provided one extremely flexible but complex solution, and one extremely simple. The 'separate temp/brightness' one was the complex one.

> Should the offset be linear or multiplicative? It seems in your code example it's multiplicative, but I'm not sure if that's desired. With multiplicative offset, different actual values of brightness/temperature will be affected in different ways.

The code example was for the simple approach - in it's simplest form I can think of, this issue could be closed by means of the steps in the TL;DR. I bet you could code that in under an hour... But, of course this would not be suitable for everyone. Undoubtedly, the future would hold feature requests for other ways to handle it. And we are already seeing that now. As we can see from the desired behaviour in this thread, things can very quickly spiral out of control toward the 'super complex' solution I gave as an example.

I am reminded of #131 here. I think that in the end, the refactor required for that issue, would also serve this one, and vice-versa. Sadly, I think that just like a simple 'shift the sunset times', a simple 'hotkeys to dim display', once one considers implementation, quickly becomes a re-write :(


> Additionally, what should happen to manual offset if the user opens settings and starts moving the actual brightness/temperature sliders around? I'd guess the offset should be reset then, but that would add complexity.

This is a great example of the kind of thing that can quickly become very complex. I can really see a need for both behaviours.


 As @fabiorfreitas said:
>  There are a lot of equally reasonable approaches:

And that really comes down to use-case. Like @anderrom says:

> The only reason I need this is for when I turn off the lights in my room.

Where as for me, I am trying to regulate melanopsin levels for medical reasons, and my post above links other examples of desired behaviour each different based on use-case.

And, being programmers, that is how this thread started: @Tyrrrz and I asking, "why would you want that?"  - because use case defines the app's behaviour, in the end. Since a fully flexible solution is in the pipeline for future releases, I think that for now, a 'quick and dirty' approach would be best, to give some functionality to the users but also to not expend too much development effort that would detract from that larger work.

**Comment by Tyrrrz** [comment]
@anderrom 

> I was thinking in a independent brightness settings, no need to have nighttime brightness or daytime brightness, so it would be much simpler to implement, and that is all usability I would ever need from this.

In that case, this is essentially the "option 2" I was talking about earlier.

**Comment by xcasxcursex** [comment]
I think that for @anderrom (please correct me if I have misunderstood here!) his requirements could be easily met with option 1, the simple hotkey-modified multiplier, combined with setting daytime and nightime brightness to 100%. This way, his brightness would only change when he presses the keys to change it.

**Comment by anderrom** [comment]
Both options would met my needs, I think option 2 would be simpler to implement, but that's up to you guys.

**Comment by Tyrrrz** [comment]
Change is in master, you're welcome to test from CI artifacts.
